File: examples/vision/detection/paddledetection/python/infer_pptinypose.py
import fastdeploy as fd
import cv2
import os


def parse_arguments():
    import argparse
    import ast
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--tinypose_model_dir",
        required=True,
        help="path of paddletinypose model directory")
    parser.add_argument(
        "--det_model_dir", help="path of paddledetection model directory")
    parser.add_argument(
        "--image", required=True, help="path of test image file.")
    parser.add_argument(
        "--device",
        type=str,
        default='cpu',
        help="type of inference device, support 'cpu' or 'gpu'.")
    parser.add_argument(
        "--use_trt",
        type=ast.literal_eval,
        default=False,
        help="wether to use tensorrt.")
    return parser.parse_args()


def build_picodet_option(args):
    option = fd.RuntimeOption()

    if args.device.lower() == "gpu":
        option.use_gpu()

    if args.use_trt:
        option.use_trt_backend()
        option.set_trt_input_shape("image", [1, 3, 320, 320])
        option.set_trt_input_shape("scale_factor", [1, 2])
    return option


def build_tinypose_option(args):
    option = fd.RuntimeOption()

    if args.device.lower() == "gpu":
        option.use_gpu()

    if args.use_trt:
        option.use_trt_backend()
        option.set_trt_input_shape("image", [1, 3, 128, 96])
    return option


args = parse_arguments()
det_result = None
if args.det_model_dir:
    picodet_model_file = os.path.join(args.det_model_dir, "model.pdmodel")
    picodet_params_file = os.path.join(args.det_model_dir, "model.pdiparams")
    picodet_config_file = os.path.join(args.det_model_dir, "infer_cfg.yml")

    # 配置runtime，加载模型
    runtime_option = build_picodet_option(args)
    det_model = fd.vision.detection.PicoDet(
        picodet_model_file,
        picodet_params_file,
        picodet_config_file,
        runtime_option=runtime_option)

    # 预测图片检测结果
    im = cv2.imread(args.image)
    det_result = det_model.predict(im.copy())
    print("PicoDet Result:\n", det_result)

    # 预测结果可视化
    det_vis_im = fd.vision.vis_detection(im, det_result, score_threshold=0.5)
    cv2.imwrite("det_visualized_result.jpg", det_vis_im)
    print("Detection visualized result save in ./det_visualized_result.jpg")

tinypose_model_file = os.path.join(args.tinypose_model_dir, "model.pdmodel")
tinypose_params_file = os.path.join(args.tinypose_model_dir, "model.pdiparams")
tinypose_config_file = os.path.join(args.tinypose_model_dir, "infer_cfg.yml")
# 配置runtime，加载模型
runtime_option = build_tinypose_option(args)
tinypose_model = fd.vision.detection.PPTINYPOSE(
    tinypose_model_file,
    tinypose_params_file,
    tinypose_config_file,
    runtime_option=runtime_option)
# 预测图片检测结果
im = cv2.imread(args.image)
tinypose_result = tinypose_model.predict(im.copy(), det_result)
print("Paddle TinyPose Result:\n", tinypose_result)

# 预测结果可视化
vis_im = fd.vision.vis_keypoint_detection(
    im, tinypose_result, conf_threshold=0.5)
cv2.imwrite("visualized_result.jpg", vis_im)
print("TinyPose visualized result save in ./visualized_result.jpg")

File: examples/vision/keypointdetection/cpp/infer_pptinypose.cc
// Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "fastdeploy/vision.h"

#ifdef WIN32
const char sep = '\\';
#else
const char sep = '/';
#endif

void CpuInfer(const std::string& tinypose_model_dir,
              const std::string& image_file) {
  auto tinypose_model_file = tinypose_model_dir + sep + "model.pdmodel";
  auto tinypose_params_file = tinypose_model_dir + sep + "model.pdiparams";
  auto tinypose_config_file = tinypose_model_dir + sep + "infer_cfg.yml";
  auto tinypose_model = fastdeploy::vision::keypointdetection::PPTinyPose(
      tinypose_model_file, tinypose_params_file, tinypose_config_file);
  if (!tinypose_model.Initialized()) {
    std::cerr << "TinyPose Model Failed to initialize." << std::endl;
    return;
  }

  auto im = cv::imread(image_file);
  cv::Mat tinypose_im_bak = im.clone();
  fastdeploy::vision::KeyPointDetectionResult res;
  if (!tinypose_model.Predict(&tinypose_im_bak, &res)) {
    std::cerr << "TinyPose Prediction Failed." << std::endl;
    return;
  } else {
    std::cout << "TinyPose Prediction Done!" << std::endl;
  }
  // 输出预测框结果
  std::cout << res.Str() << std::endl;

  // 可视化预测结果
  auto tinypose_vis_im =
      fastdeploy::vision::Visualize::VisKeypointDetection(im, res, 0.5);
  cv::imwrite("tinypose_vis_result.jpg", tinypose_vis_im);
  std::cout << "TinyPose visualized result saved in ./tinypose_vis_result.jpg"
            << std::endl;
}

void GpuInfer(const std::string& tinypose_model_dir,
              const std::string& image_file) {
  auto option = fastdeploy::RuntimeOption();
  option.UseGpu();

  auto tinypose_model_file = tinypose_model_dir + sep + "model.pdmodel";
  auto tinypose_params_file = tinypose_model_dir + sep + "model.pdiparams";
  auto tinypose_config_file = tinypose_model_dir + sep + "infer_cfg.yml";
  auto tinypose_model = fastdeploy::vision::keypointdetection::PPTinyPose(
      tinypose_model_file, tinypose_params_file, tinypose_config_file, option);
  if (!tinypose_model.Initialized()) {
    std::cerr << "TinyPose Model Failed to initialize." << std::endl;
    return;
  }

  auto im = cv::imread(image_file);
  auto tinypose_im_bak = im.clone();
  fastdeploy::vision::KeyPointDetectionResult res;
  if (!tinypose_model.Predict(&tinypose_im_bak, &res)) {
    std::cerr << "TinyPose Prediction Failed." << std::endl;
    return;
  } else {
    std::cout << "TinyPose Prediction Done!" << std::endl;
  }
  // 输出预测框结果
  std::cout << res.Str() << std::endl;

  // 可视化预测结果
  auto tinypose_vis_im =
      fastdeploy::vision::Visualize::VisKeypointDetection(im, res, 0.5);
  cv::imwrite("tinypose_vis_result.jpg", tinypose_vis_im);
  std::cout << "TinyPose visualized result saved in ./tinypose_vis_result.jpg"
            << std::endl;
}

void TrtInfer(const std::string& tinypose_model_dir,
              const std::string& image_file) {
  auto tinypose_model_file = tinypose_model_dir + sep + "model.pdmodel";
  auto tinypose_params_file = tinypose_model_dir + sep + "model.pdiparams";
  auto tinypose_config_file = tinypose_model_dir + sep + "infer_cfg.yml";
  auto tinypose_option = fastdeploy::RuntimeOption();
  tinypose_option.UseGpu();
  tinypose_option.UseTrtBackend();
  auto tinypose_model = fastdeploy::vision::keypointdetection::PPTinyPose(
      tinypose_model_file, tinypose_params_file, tinypose_config_file,
      tinypose_option);
  if (!tinypose_model.Initialized()) {
    std::cerr << "TinyPose Model Failed to initialize." << std::endl;
    return;
  }

  auto im = cv::imread(image_file);
  cv::Mat tinypose_im_bak = im.clone();
  fastdeploy::vision::KeyPointDetectionResult res;
  if (!tinypose_model.Predict(&tinypose_im_bak, &res)) {
    std::cerr << "TinyPose Prediction Failed." << std::endl;
    return;
  } else {
    std::cout << "TinyPose Prediction Done!" << std::endl;
  }
  // 输出预测框结果
  std::cout << res.Str() << std::endl;

  // 可视化预测结果
  auto tinypose_vis_im =
      fastdeploy::vision::Visualize::VisKeypointDetection(im, res, 0.5);
  cv::imwrite("tinypose_vis_result.jpg", tinypose_vis_im);
  std::cout << "TinyPose visualized result saved in ./tinypose_vis_result.jpg"
            << std::endl;
}

int main(int argc, char* argv[]) {
  if (argc < 4) {
    std::cout << "Usage: infer_demo path/to/pptinypose_model_dir path/to/image "
                 "run_option, "
                 "e.g ./infer_model ./pptinypose_model_dir ./test.jpeg 0"
              << std::endl;
    std::cout << "The data type of run_option is int, 0: run with cpu; 1: run "
                 "with gpu; 2: run with gpu and use tensorrt backend."
              << std::endl;
    return -1;
  }

  if (std::atoi(argv[3]) == 0) {
    CpuInfer(argv[1], argv[2]);
  } else if (std::atoi(argv[3]) == 1) {
    GpuInfer(argv[1], argv[2]);
  } else if (std::atoi(argv[3]) == 2) {
    TrtInfer(argv[1], argv[2]);
  }
  return 0;
}
